File: core/src/main/java/com/cangwang/core/cwmodule/ex/CWModuleExFactory.java
package com.cangwang.core.cwmodule.ex;

import com.cangwang.model.IModuleFactory;

/**
 * Created by cangwang on 2017/6/16.
 */

public class CWModuleExFactory {
    private static final String FACTORY_PATH = "com.cangwang.core.ModuleCenterFactory";

    private static IModuleFactory instance;

    public static CWAbsExModule newModuleInstance(String name){
        if (name ==null || name.equals("")){
            return null;
        }

        try{
            Class<? extends CWAbsExModule> moduleClzz = (Class<? extends CWAbsExModule>) Class.forName(name);
            if (moduleClzz !=null){
                CWAbsExModule instance = (CWAbsExModule)moduleClzz.newInstance();
                return instance;
            }
            return null;
        }catch (ClassNotFoundException e){
            e.printStackTrace();
        }catch (InstantiationException e){
            e.printStackTrace();
        }catch (IllegalAccessException e){
            e.printStackTrace();
        }
        return null;
    }

    public static IModuleFactory getInstance(){
        if (instance == null) {
            try {
                Class<? extends IModuleFactory> factoryClazz = (Class<? extends IModuleFactory>) Class.forName(FACTORY_PATH);
                if (factoryClazz != null) {
                    instance = factoryClazz.newInstance();
                    return instance;
                }
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            } catch (InstantiationException e) {
                e.printStackTrace();
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            }
            return null;
        }else
            return instance;
    }
}

File: core/src/main/java/com/cangwang/core/cwmodule/ex/CWModuleExFactory.kt
package com.cangwang.core.cwmodule.ex

import com.cangwang.model.IModuleFactory

/**
 * Created by cangwang on 2017/6/16.
 */
object CWModuleExFactory {
    private const val FACTORY_PATH = "com.cangwang.core.ModuleCenterFactory"
    var instance: IModuleFactory? = null
        get() = if (field == null) {
            try {
                val factoryClazz = Class.forName(FACTORY_PATH) as Class<out IModuleFactory>
                if (factoryClazz != null) {
                    field = factoryClazz.newInstance()
                    field
                } else {
                    null
                }
            } catch (e: ClassNotFoundException) {
                e.printStackTrace()
                null
            } catch (e: InstantiationException) {
                e.printStackTrace()
                null
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
                null
            }
        } else field
        private set

    fun newModuleInstance(name: String?): CWAbsExModule? {
        if (name == null || name == "") {
            return null
        }
        try {
            val moduleClzz = Class.forName(name) as Class<out CWAbsExModule>
            return if (moduleClzz != null) {
                moduleClzz.newInstance() as CWAbsExModule
            } else null
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
        return null
    }
}